iceoryx_hoofs/container/include/iox/detail/vector.inl

```diff
@@ -359,13 +408,26 @@ inline bool vector<T, Capacity>::erase(iterator position) noexcept
         // AXIVION Next Line AutosarC++19_03-M5.0.9 : False positive. Pointer arithmetic occurs here.
         uint64_t index{static_cast<uint64_t>(position - begin())};
         uint64_t n{index};
-        while ((n + 1U) < size())
+        if constexpr (std::is_trivially_copyable<T>::value)
+        {
+            if constexpr (!(std::is_trivially_destructible<T>::value))
+            {
+                at_unchecked(n).~T();
+            }
```

Comment:
I do not clearly understand your question.
Do you want to know if the ```if constexpr``` will remove the ```at_unchecked``` call?

Reply:
No, I was just curios if the constexpr has a real effect. It might be that the compiler would optimize the dtor call away anyway, even without the constexpr.

If you are also curious you could try it on godbolt. If not, we can merge this PR.

Reply:
I created a small [test](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QAJgCsfAQQCqAZ0wAFAB6cADLwBWILqVpMGoZAFIJAIQuXSy%2BsgJ4BlRugDCqWgFcWDEADMAGykbgAyeAyYAHK%2BAEaYxCAA7KQADqiKhM4Mnj5%2BgSEZWU4CEVGxLAlJqfaYjjlCBEzEBHm%2B/sF2mA6lDE0tBOUx8Ykpds2t7QVdipNDkSNVY8kAlHao3sTI7BwWAZHIPlgA1GYB7s5zxJis59hmmgCCj08EmCxpRu9nFwQAnmlGKxMCcACqAzD3V5HJiKRQnABq9SIxDMyWszxO2JOaW8cVoeGQIFeOLJSJRJAgq3RlhOWQAXqDzgARE6ac509Es0nk7EAN1QeHQuO8igQAH04kxkABrCAQoEnflibyYGkY3l8snoJjNMxSSyMqFSNms5WqqEBTFPbXk402R3WrXk7kusmC4W44iRAjU2nu7X8YgQX0nPC/Nkc63h37uel4Jmc8OOqwam12u1zdAgFCbAjndyF3X6w14A08i6FiwSA3uBg1zmBvlurHa1u27UAP2Rjip6eb5ODocEsfN0bpEcLCaTMfLVhsA7bmfJJaYBss5dNADou4rMP7ncvMx27aecWkfSr3iTj2T9yc1xuuJoOaam3ecWGHUeXslKzarxhiwTCRP6mrHr2qKFlQtCoHq9zKvUH6dgK9Tbni4pSjK8ovtumgAGI0r%2BHroZhkrSnKEASJo%2BFEShLr8mRPqCIegHHjcBBbAw7IMf%2BHDrLQnBSLw/gcNopCoJwRYLlY9KbNszISAEPCkAQWgCessogFIEjbsEAAsBkAJxcGZRkGQAHJZEiGJwBmiRpkmcLwiggJoakaescCwDAiB5p8dCJOQlBoIF9BJMAxkGXwdDvMQbkQHETlxJELT/JwqmpcwxD/AA8nEOgoplvBhWwgh5QwtAZeJvBYHE3jAO4Yi0G53B1R8xjAOItWkPgNwNExbUSZgaj1N4N69b6PROYScTEOlnhYCVak%2BiwXl8EYwCKIieCYAA7nlQJiap/CCCIYjsFwMVnQoyjqE5egGEYJgoKmthzW5sDMGwIALSYmCKPyASkExSTicEmmSWkfRtQAtNmrLmLJlhcMkJxw3lqmoGDPpYF91LdL0OSuAwHheB0%2BihGTwyVNUVPFNkAjTP4BiM30tOjEkBh1A0AgDFMFMFDzPQoo08yc8s3MTIMLNU3MgyS/TXDrIoCk7PognCY5vVSRwJxqJZQRw0EBknMAyDICcxnbmbEC4IQJBnMpKu8OptWrFp%2BivnZHAOaQ61G9uxnGVIyQSEEYcvpZySm6QYkSXrrnuZ5HukD5/n5niBAhRAmD4KiwpU7dF3iNdsiCEoqgaL1T2kPtC1pCVWscCJ8dOXreUTdnJyoFQBtGybZsW1bNt2544WJM7KmrG7XnrAgtxYEkhNCX7vDrVwAT4ZoBkSMkAR75ZpnKckMUJ7wSd2Cn7vaJ7pDaQZyTbkEARSFImjWZZ79SGZlm%2BwEHWicXKpzvi3CQQDL4gNvlDMGWQXAGSAA%3D%3D) and with ```-O1``` seems that the compiler remove the dtor call. Maybe is because is a simple example? I don't know!

Reply:
No, I'm quite sure it's not related o the simple example. This is one of the first optimizations compilers do. The compiler is allowed to do anything that cannot be observed by the user and just don't call anything that does not have any effect. Since trivially destructible types do not have an effect when the dtor is called the compiler is happy to elide the call.

I think the dtor can be callen unconditionally here and in `pop_back`. Just a small simplification.

In general it is always good to check whether an explicit optimization is necessary since compilers are quite good nowadays to detect common pattern and optimize them. Often it is even counterproductive to try to be smarter than the compiler because this can lead to situations where the compiler isn't able to detect a pattern anymore and the result is slower than the naive code. The rule of thumb is to always measure before optimizing :)

Reply:
I partially agree with you, I used different compilers and not all are smarter.

I could leave the dcot call if you want, tell me what to do :)

Reply:
In this case let's keep it.